plugins/inputs/opensearch_query/opensearch_query.go

```diff
@@ -0,0 +1,326 @@
+	defer cancel()
+
+	searchResult, err := o.runAggregationQuery(ctx, aggregation)
+	if err != nil {
+		return err
+	}
+
+	return searchResult.GetMetrics(acc, aggregation.MeasurementName)
+}
+
+func init() {
+	inputs.Add("opensearch_query", func() telegraf.Input {
+		return &OpensearchQuery{
+			Timeout:             config.Duration(time.Second * 5),
+			HealthCheckInterval: config.Duration(time.Second * 10),
+		}
+	})
+}
+
+func (o *OpensearchQuery) runAggregationQuery(ctx context.Context, aggregation osAggregation) (*AggregationResponse, error) {
+	now := time.Now().UTC()
+	from := now.Add(time.Duration(-aggregation.QueryPeriod))
+	filterQuery := aggregation.FilterQuery
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	aq := &Query{
+		Size:         0,
+		Aggregations: aggregation.aggregation,
+		Query:        nil,
+	}
+
+	boolQuery := &BoolQuery{
+		FilterQueryString: filterQuery,
+		TimestampField:    aggregation.DateField,
+		TimeRangeFrom:     from,
+		TimeRangeTo:       now,
+		DateFieldFormat:   aggregation.DateFieldFormat,
+	}
+
+	aq.Query = boolQuery
+	req, err := json.Marshal(aq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %s", err)
+	}
+	o.Log.Debugf("{\"body\": %s}", string(req))
+
+	searchRequest := &opensearchapi.SearchRequest{
+		Body:    strings.NewReader(string(req)),
+		Index:   []string{aggregation.Index},
+		Timeout: time.Duration(o.Timeout),
+	}
+
+	resp, err := searchRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+	if resp.IsError() {
+		return nil, fmt.Errorf("Opensearch SearchRequest failure: [%d] %s", resp.StatusCode, resp.Status())
+	}
+	defer resp.Body.Close()
+
+	var searchResult AggregationResponse
+
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&searchResult)
+	if err != nil {
+		return nil, err
+	}
+
+	return &searchResult, nil
+}
+
+// getMetricFields function returns a map of fields and field types on Elasticsearch that matches field.MetricFields
+func (o *OpensearchQuery) getMetricFields(ctx context.Context, aggregation osAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+	fieldMappingRequest := opensearchapi.IndicesGetFieldMappingRequest{
+		Index:  []string{aggregation.Index},
+		Fields: aggregation.MetricFields,
+	}
+
+	response, err := fieldMappingRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// Bad request; move on
+	if response.StatusCode != 200 {
+		return mapMetricFields, nil
+	}
+
+	var m mapping
+	decoder := json.NewDecoder(response.Body)
+	err = decoder.Decode(&m)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mm := range m {
+		for _, f := range aggregation.MetricFields {
+			if _, ok := mm.Mappings[f]; ok {
+				mapMetricFields[f] = mm.Mappings[f].Mapping[f].Type
+			}
+		}
+	}
+
+	return mapMetricFields, nil
+}
+
+func (aggregation *osAggregation) buildAggregationQuery() error {
+	var agg AggregationRequest
+	agg = &MetricAggregationRequest{}
```

Comment:
```suggestion
	agg := &MetricAggregationRequest{}
```

Reply:
This should stay as-is.  `AggregationRequest` is an interface used to build the query to the cluster.  If there are tags, then the final product will be of type `BucketAggregationRequest` with a type `MetricAggregationRequest` nested below, but if there are not, it will be of type `MetricAggregationRequest`.

Reply:
Oh I completely missed this. You are right, leave as-is.